File: python/mmd_litellm/alignment.py
from __future__ import annotations

from dataclasses import dataclass

from .schemas import CandidateClaim


@dataclass(frozen=True)
class PairSupport:
    left_claim_id: str
    right_claim_id: str
    support: int
    cannot_link: bool


def _pair_key(left: str, right: str) -> tuple[str, str]:
    return tuple(sorted((left, right)))  # type: ignore[return-value]


def deterministic_complete_link(
    claim_ids: list[str], pair_support: list[PairSupport], minimum_support: int
) -> tuple[list[list[str]], list[dict[str, object]]]:
    if minimum_support < 1:
        raise ValueError("minimum_support must be positive")
    if len(set(claim_ids)) != len(claim_ids):
        raise ValueError("claim_ids must be unique")
    support = {
        _pair_key(pair.left_claim_id, pair.right_claim_id): pair
        for pair in pair_support
    }
    clusters = [[claim_id] for claim_id in sorted(claim_ids)]
    decisions: list[dict[str, object]] = []
    while True:
        merged = False
        cluster_pairs = [
            (left_index, right_index)
            for left_index in range(len(clusters))
            for right_index in range(left_index + 1, len(clusters))
        ]
        cluster_pairs.sort(key=lambda item: (clusters[item[0]], clusters[item[1]]))
        for left_index, right_index in cluster_pairs:
            left = clusters[left_index]
            right = clusters[right_index]
            cross = [support.get(_pair_key(a, b)) for a in left for b in right]
            cannot_link = any(pair is not None and pair.cannot_link for pair in cross)
            fully_supported = all(
                pair is not None and pair.support >= minimum_support for pair in cross
            )
            if cannot_link or not fully_supported:
                decisions.append(
                    {
                        "left": list(left),
                        "right": list(right),
                        "action": "reject",
                        "reason": "cannot_link" if cannot_link else "insufficient_support",
                    }
                )
                continue
            decisions.append(
                {
                    "left": list(left),
                    "right": list(right),
                    "action": "merge",
                    "reason": "complete_link_supported",
                }
            )
            clusters = [
                cluster
                for index, cluster in enumerate(clusters)
                if index not in (left_index, right_index)
            ]
            clusters.append(sorted([*left, *right]))
            clusters.sort()
            merged = True
            break
        if not merged:
            return clusters, decisions


def candidates_from_clusters(
    *, run_id: str, topic_id: str | None, claims: list[object], clusters: list[list[str]]
) -> list[CandidateClaim]:
    by_id = {getattr(claim, "claim_id"): claim for claim in claims}
    scope = topic_id or "root"
    candidates: list[CandidateClaim] = []
    for index, source_ids in enumerate(sorted([sorted(cluster) for cluster in clusters])):
        representative = by_id[source_ids[0]]
        candidates.append(
            CandidateClaim(
                candidate_id=f"{run_id}::{scope}::candidate::{index:03d}",
                text=getattr(representative, "text"),
                source_claim_ids=source_ids,
                topic_id=topic_id,
                notes="deterministic complete-link cluster",
            )
        )
    return candidates

File: packages/protocol/src/alignment.ts
import type { CandidateClaim } from "./schemas/normalize.js";

export interface AlignmentClaim {
  claim_id: string;
  text: string;
  model_id: string;
  topic_id?: string;
}

export interface PairSupport {
  left_claim_id: string;
  right_claim_id: string;
  support: number;
  cannot_link: boolean;
}

export interface ClusterDecision {
  left: string[];
  right: string[];
  action: "merge" | "reject";
  reason: "complete_link_supported" | "insufficient_support" | "cannot_link";
}

export interface CompleteLinkResult {
  clusters: string[][];
  decisions: ClusterDecision[];
}

function pairKey(left: string, right: string): string {
  return [left, right].sort().join("\u0000");
}

function clusterKey(cluster: string[]): string {
  return [...cluster].sort().join("\u0001");
}

export function deterministicCompleteLink(params: {
  claimIds: string[];
  pairSupport: PairSupport[];
  minimumSupport: number;
}): CompleteLinkResult {
  if (!Number.isInteger(params.minimumSupport) || params.minimumSupport < 1) {
    throw new Error("minimumSupport must be a positive integer");
  }
  const unique = [...new Set(params.claimIds)].sort();
  if (unique.length !== params.claimIds.length) {
    throw new Error("claimIds must be unique");
  }
  const support = new Map(
    params.pairSupport.map((pair) => [
      pairKey(pair.left_claim_id, pair.right_claim_id),
      pair,
    ])
  );
  let clusters = unique.map((id) => [id]);
  const decisions: ClusterDecision[] = [];

  while (true) {
    const pairs: Array<[number, number]> = [];
    for (let left = 0; left < clusters.length; left += 1) {
      for (let right = left + 1; right < clusters.length; right += 1) {
        pairs.push([left, right]);
      }
    }
    pairs.sort((a, b) => {
      const ak = `${clusterKey(clusters[a[0]])}\u0002${clusterKey(clusters[a[1]])}`;
      const bk = `${clusterKey(clusters[b[0]])}\u0002${clusterKey(clusters[b[1]])}`;
      return ak.localeCompare(bk);
    });

    let merged = false;
    for (const [leftIndex, rightIndex] of pairs) {
      const left = clusters[leftIndex];
      const right = clusters[rightIndex];
      const crossPairs = left.flatMap((leftId) =>
        right.map((rightId) => support.get(pairKey(leftId, rightId)))
      );
      const hasCannotLink = crossPairs.some((pair) => pair?.cannot_link);
      const fullySupported = crossPairs.every(
        (pair) => pair !== undefined && pair.support >= params.minimumSupport
      );
      if (hasCannotLink || !fullySupported) {
        decisions.push({
          left: [...left],
          right: [...right],
          action: "reject",
          reason: hasCannotLink ? "cannot_link" : "insufficient_support",
        });
        continue;
      }
      decisions.push({
        left: [...left],
        right: [...right],
        action: "merge",
        reason: "complete_link_supported",
      });
      const next = clusters.filter(
        (_cluster, index) => index !== leftIndex && index !== rightIndex
      );
      next.push([...left, ...right].sort());
      clusters = next.sort((a, b) => clusterKey(a).localeCompare(clusterKey(b)));
      merged = true;
      break;
    }
    if (!merged) break;
  }

  return { clusters, decisions };
}

export function candidatesFromClusters(params: {
  runId: string;
  topicId?: string;
  claims: AlignmentClaim[];
  clusters: string[][];
}): CandidateClaim[] {
  const byId = new Map(params.claims.map((claim) => [claim.claim_id, claim]));
  const scope = params.topicId ?? "root";
  return [...params.clusters]
    .map((cluster) => [...cluster].sort())
    .sort((a, b) => clusterKey(a).localeCompare(clusterKey(b)))
    .map((sourceIds, index) => {
      const representative = byId.get(sourceIds[0]);
      if (!representative) throw new Error(`unknown claim id: ${sourceIds[0]}`);
      return {
        candidate_id: `${params.runId}::${scope}::candidate::${String(index).padStart(3, "0")}`,
        text: representative.text,
        source_claim_ids: sourceIds,
        topic_id: params.topicId,
        notes: "deterministic complete-link cluster",
      };
    });
}
